### ISSUE TITLE
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._init_func at 0x7fadaffa7d90> could not be transformed and will be executed as-is.

### ISSUE THREAD

**Issue by Bengt** [tensorflow/autograph#4]
The full error message:

```
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._init_func at 0x7f8069decd90> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._remote_init_func at 0x7f8068094268> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._next_func at 0x7f80680949d8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._remote_next_func at 0x7f7df0609158> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._finalize_func at 0x7f7df0609bf8> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
WARNING:tensorflow:Entity <function _CopyToDeviceDataset.__init__.<locals>._remote_finalize_func at 0x7f7df0609ea0> could not be transformed and will be executed as-is. Please report this to the AutoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: Bad argument number for Name: 3, expecting 4
```

**Comment by Bengt** [comment]
There are also logs for other function names. The full list:

-   `_init_func`
-   `_remote_init_func`
-   `_next_func`
-   `_remote_next_func`
-   `_finalize_func`
-   `_remote_finalize_func`

**Comment by Bengt** [comment]
As a workaround, I can comment out this part of my data set transformation code:

```
    dataset_training = dataset_training.apply(
        transformation_func=tf.data.experimental.prefetch_to_device(
            device='/gpu:0',
            buffer_size=1,
        )
    )
```

This makes all warnings disappear from the log.

**Comment by mdanatg** [comment]
@Bengt this might be caused by https://github.com/tensorflow/autograph/issues/1. Could you try to see if running this fixes it:

`pip install --user gast==0.2.2`

**Comment by Bengt** [comment]
Hi, @mdanatg! Thanks for the quick reply. Yes, downgrading gast actually solved my issue. I thought I had done that already, but I didn't. Prefetching to a device works now and actually gave me a measurable speedup.

The documentation states that "this transformation must be the final dataset in the input pipeline." I use two GPUs and thus call the map twice per dataset with the `device` parameter set to each GPU once. Is this the correct usage?

```python
    for device in ('/gpu:0', '/gpu:1'):
        dataset_training = dataset_training.apply(
            transformation_func=tf.data.experimental.prefetch_to_device(
                device=device, buffer_size=tf.data.experimental.AUTOTUNE,
            )
        )
        dataset_validation = dataset_validation.apply(
            transformation_func=tf.data.experimental.prefetch_to_device(
                device=device, buffer_size=tf.data.experimental.AUTOTUNE,
            )
        )
        dataset_testing = dataset_testing.apply(
            transformation_func=tf.data.experimental.prefetch_to_device(
                device=device, buffer_size=tf.data.experimental.AUTOTUNE,
            )
        )
```

**Comment by mdanatg** [comment]
@jsimsa for the dataset transformation question

@Bengt good to hear! We're working on permanently fixing that in TF 2.2.

**Comment by jsimsa** [comment]
Regarding your tf.data usage, you program will create separate instances of the dataset to feed each GPU. Is that your intention?

If not, I recommend that you take a look at [tf.distribute.MirroredStrategy](https://www.tensorflow.org/api_docs/python/tf/distribute/MirroredStrategy?version=stable), which can be used to create a single instance of the dataset and prefetch its data to the GPU.

**Comment by Bengt** [comment]
Thanks for the advice,  @jsimsa! I do not mind fetching the whole dataset to all GPUs since my samples are small and VRAM thus plenty. Yes, I already use the `MirroredStrategy` to create the model. However, I am affected by a bug with TensorFlow, where `prefetch` does not work, but `prefetch_to_device` does: https://github.com/tensorflow/tensorflow/issues/28782

**Comment by jsimsa** [comment]
That bug should be fixed in TF 2.1 which was released on January 8th. Could you try with the latest version?

**Comment by Bengt** [comment]
Yes, I can use tensorflow-rocm==2.1.0-rc1 from PyPI, that solves this issue and I run into this one:

https://github.com/tensorflow/tensorflow/issues/35100
https://github.com/tensorflow/tensorflow/issues/24570

I guess I will have to wait at least for the final 2.1 for a proper fix.

**Comment by Bengt** [comment]
For reference, using the broken `prefetch` (at 14 seconds per epoch) like this is about 14 times slower than `prefetch_to_device` (at 1 seconds per epoch) for me:

```python
    dataset_training = dataset_training.prefetch\
        (buffer_size=tf.data.experimental.AUTOTUNE)
    dataset_validation = dataset_validation.prefetch\
        (buffer_size=tf.data.experimental.AUTOTUNE)
```

**Comment by jsimsa** [comment]
The `GeneratorDataset` warning can be safely ignored. Unfortunately, it looks like my fix (https://github.com/tensorflow/tensorflow/commit/b6edd34c5858ab0ab4380da774e7e2fd81a92da0) that silences the warning didn't make it into TF 2.1.

**Comment by mdanatg** [comment]
I believe all the issues we've talked about have been addressed, although some fixes will only land in 2.2. Feel free to reopen if we missed anything.
